File: internal/wire/decoder.go
// Copyright 2026 The go-hbase Authors
// SPDX-License-Identifier: Apache-2.0

package wire

import (
	"encoding/binary"
	"fmt"
	"io"
)

type Decoder struct {
	r            io.Reader
	pending      map[uint64]*reassembly
	maxFrame     int
	pendingBytes int
}

type reassembly struct {
	typ      Type
	regionID uint32
	hookID   uint8
	total    uint32
	chunks   [][]byte // index = chunk_idx; nil until that chunk arrives
	received uint32
	size     int
}

func NewDecoder(r io.Reader) *Decoder {
	return &Decoder{
		r:        r,
		pending:  make(map[uint64]*reassembly),
		maxFrame: MaxFrameSize,
	}
}

func (d *Decoder) Decode() (*Message, error) {
	for {
		msg, err := d.readChunk()
		if err != nil {
			return nil, err
		}
		if msg != nil {
			return msg, nil
		}
	}
}

func (d *Decoder) readChunk() (*Message, error) {
	var lenBuf [4]byte
	if _, err := io.ReadFull(d.r, lenBuf[:]); err != nil {
		return nil, err
	}
	l := binary.BigEndian.Uint32(lenBuf[:])

	if l < headerSize || uint64(l) > uint64(d.maxFrame)-4 {
		return nil, fmt.Errorf("%w: %d", ErrFrameTooLarge, l)
	}

	buf := make([]byte, l)
	if _, err := io.ReadFull(d.r, buf); err != nil {
		return nil, err
	}

	typ := Type(buf[0])
	if !typ.Valid() {
		return nil, fmt.Errorf("%w: %d", ErrUnknownType, typ)
	}
	reqID := binary.BigEndian.Uint64(buf[1:9])
	regionID := binary.BigEndian.Uint32(buf[9:13])
	hookID := buf[13]
	// buf[14] is chunk_flags, reserved.
	chunkIdx := binary.BigEndian.Uint32(buf[15:19])
	chunkTotal := binary.BigEndian.Uint32(buf[19:23])
	payload := buf[23:]

	if chunkTotal == 0 || chunkIdx >= chunkTotal {
		return nil, fmt.Errorf("%w: idx=%d total=%d", ErrInvalidChunk, chunkIdx, chunkTotal)
	}
	if chunkTotal > MaxChunks {
		return nil, fmt.Errorf("%w: %d > %d", ErrTooManyChunks, chunkTotal, MaxChunks)
	}

	if chunkTotal == 1 {
		out := append([]byte(nil), payload...)
		return &Message{
			Type: typ, ReqID: reqID, RegionID: regionID, HookID: hookID,
			Payload: out,
		}, nil
	}

	if typ.isControl() {
		return nil, fmt.Errorf("%w: type=%d", ErrControlMultiChunk, typ)
	}
	if reqID == 0 {
		return nil, fmt.Errorf("%w: multi-chunk requires non-zero req_id", ErrInvalidChunk)
	}

	re, ok := d.pending[reqID]
	if !ok {
		if len(d.pending) >= MaxPendingReassemblies {
			return nil, fmt.Errorf("%w: %d", ErrTooManyPending, len(d.pending))
		}
		re = &reassembly{
			typ: typ, regionID: regionID, hookID: hookID, total: chunkTotal,
			chunks: make([][]byte, chunkTotal),
		}
		d.pending[reqID] = re
	} else if re.typ != typ || re.regionID != regionID || re.hookID != hookID || re.total != chunkTotal {
		return nil, fmt.Errorf("%w: header drift on req_id %d", ErrInvalidChunk, reqID)
	}
	if re.chunks[chunkIdx] != nil {
		return nil, fmt.Errorf("%w: duplicate chunk_idx %d for req_id %d", ErrInvalidChunk, chunkIdx, reqID)
	}

	if d.pendingBytes+len(payload) > MaxPendingBytes {
		return nil, fmt.Errorf("%w: %d + %d > %d", ErrTooManyPendingBytes, d.pendingBytes, len(payload), MaxPendingBytes)
	}
	chunkCopy := append([]byte(nil), payload...)
	re.chunks[chunkIdx] = chunkCopy
	re.received++
	re.size += len(chunkCopy)
	d.pendingBytes += len(chunkCopy)

	if re.received < re.total {
		return nil, nil
	}

	out := make([]byte, 0, re.size)
	for _, c := range re.chunks {
		out = append(out, c...)
	}
	delete(d.pending, reqID)
	d.pendingBytes -= re.size
	return &Message{
		Type: re.typ, ReqID: reqID, RegionID: re.regionID, HookID: re.hookID,
		Payload: out,
	}, nil
}

File: java/com/virogg/hbasecop/bridge/wire/Decoder.java
// Copyright 2026 The go-hbase Authors
// SPDX-License-Identifier: Apache-2.0

package com.virogg.hbasecop.bridge.wire;

import java.nio.ByteBuffer;
import java.util.HashMap;
import java.util.Map;

public final class Decoder {

  private final Map<Long, Reassembly> pending = new HashMap<>();
  private final int maxFrame = WireFormat.MAX_FRAME_SIZE;

  private int pendingBytes;

  public Message decode(ByteBuffer src) throws WireException {
    while (true) {
      if (!src.hasRemaining()) {
        return null;
      }
      Message msg = readChunk(src);
      if (msg != null) {
        return msg;
      }
    }
  }

  private Message readChunk(ByteBuffer src) throws WireException {
    if (src.remaining() < 4) {
      throw new FrameTooLargeException("truncated length prefix");
    }
    int len = src.getInt();
    if (len < WireFormat.HEADER_SIZE || len > maxFrame - 4) {
      throw new FrameTooLargeException("frame length out of range: " + len);
    }
    if (src.remaining() < len) {
      throw new FrameTooLargeException(
          "truncated frame body: need " + len + " have " + src.remaining());
    }

    byte typeByte = src.get();
    FrameType type = FrameType.fromByte(typeByte);
    if (!type.valid()) {
      throw new UnknownTypeException("type=" + Byte.toUnsignedInt(typeByte));
    }
    long reqId = src.getLong();
    int regionId = src.getInt();
    byte hookId = src.get();
    src.get(); // chunk_flags reserved
    int chunkIdx = src.getInt();
    int chunkTotal = src.getInt();
    int payloadLen = len - WireFormat.HEADER_SIZE;
    byte[] payload = new byte[payloadLen];
    src.get(payload);

    if (chunkTotal <= 0 || chunkIdx < 0 || chunkIdx >= chunkTotal) {
      throw new InvalidChunkException("idx=" + chunkIdx + " total=" + chunkTotal);
    }
    if (chunkTotal > WireFormat.MAX_CHUNKS) {
      throw new TooManyChunksException(chunkTotal + " > " + WireFormat.MAX_CHUNKS);
    }

    if (chunkTotal == 1) {
      return new Message(type, reqId, regionId, hookId, payload);
    }

    if (type.isControl()) {
      throw new ControlMultiChunkException("type=" + type);
    }
    if (reqId == 0) {
      throw new InvalidChunkException("multi-chunk requires non-zero req_id");
    }

    Reassembly re = pending.get(reqId);
    if (re == null) {
      if (pending.size() >= WireFormat.MAX_PENDING_REASSEMBLIES) {
        throw new TooManyPendingException("pending reassemblies: " + pending.size());
      }
      re = new Reassembly(type, regionId, hookId, chunkTotal);
      pending.put(reqId, re);
    } else if (re.type != type
        || re.regionId != regionId
        || re.hookId != hookId
        || re.total != chunkTotal) {
      throw new InvalidChunkException("header drift on req_id " + reqId);
    }
    if (re.chunks[chunkIdx] != null) {
      throw new InvalidChunkException("duplicate chunk_idx " + chunkIdx + " for req_id " + reqId);
    }
    if (pendingBytes + payload.length > WireFormat.MAX_PENDING_BYTES) {
      throw new TooManyPendingBytesException(
          pendingBytes + " + " + payload.length + " > " + WireFormat.MAX_PENDING_BYTES);
    }
    re.chunks[chunkIdx] = payload;
    re.received++;
    re.size += payload.length;
    pendingBytes += payload.length;

    if (re.received < re.total) {
      return null;
    }

    byte[] out = new byte[re.size];
    int cursor = 0;
    for (byte[] c : re.chunks) {
      System.arraycopy(c, 0, out, cursor, c.length);
      cursor += c.length;
    }
    pending.remove(reqId);
    pendingBytes -= re.size;
    return new Message(re.type, reqId, re.regionId, re.hookId, out);
  }

  private static final class Reassembly {
    final FrameType type;
    final int regionId;
    final byte hookId;
    final int total;
    final byte[][] chunks;
    int received;
    int size;

    Reassembly(FrameType type, int regionId, byte hookId, int total) {
      this.type = type;
      this.regionId = regionId;
      this.hookId = hookId;
      this.total = total;
      this.chunks = new byte[total][];
    }
  }
}
